File: wms/apps/android-tsd/app/src/main/java/pro/logoff/wms/tsd/sync/TsdSyncRunner.java
package pro.logoff.wms.tsd.sync;

import java.io.IOException;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import pro.logoff.wms.tsd.data.OperationOutbox;
import pro.logoff.wms.tsd.data.PendingOperation;
import pro.logoff.wms.tsd.network.TsdOperationRequest;
import pro.logoff.wms.tsd.network.TsdOperationResponse;
import pro.logoff.wms.tsd.network.TsdSyncRequest;
import pro.logoff.wms.tsd.network.WmsApi;
import retrofit2.Response;

public class TsdSyncRunner {
    private final OperationOutbox outbox;
    private final WmsApi api;
    private final String deviceId;

    public TsdSyncRunner(OperationOutbox outbox, WmsApi api, String deviceId) {
        this.outbox = outbox;
        this.api = api;
        this.deviceId = deviceId;
    }

    public TsdSyncSummary syncPending(String authorization) {
        return syncOperations(authorization, outbox.pending());
    }

    // FIX: a closed box does not wait behind unrelated batches. Drain only its scans.
    public TsdSyncSummary syncReceiptBatch(String authorization, String closeKey) {
        int sent = 0, applied = 0, rejected = 0, retried = 0;
        String message = "Нет операций для синхронизации";
        for (int page = 0; page < 21; page++) {
            List<PendingOperation> operations = outbox.pendingReceiptBatch(closeKey);
            if (operations.isEmpty()) break;
            TsdSyncSummary part = syncOperations(authorization, operations);
            sent += part.sent;
            applied += part.applied;
            rejected += part.rejected;
            retried += part.retried;
            message = part.message;
            if (part.rejected > 0 || part.retried > 0 || part.applied == 0) break;
        }
        return new TsdSyncSummary(sent, applied, rejected, retried, message);
    }

    private TsdSyncSummary syncOperations(String authorization, List<PendingOperation> operations) {
        if (operations.isEmpty()) {
            return new TsdSyncSummary(0, 0, 0, 0, "Нет операций для синхронизации");
        }

        try {
            List<TsdOperationRequest> requests = new ArrayList<>();
            for (PendingOperation operation : operations) {
                requests.add(toRequest(operation));
            }

            Response<List<TsdOperationResponse>> response = api
                .syncOperations(authorization, new TsdSyncRequest(requests, Instant.now().toString()))
                .execute();
            if (!response.isSuccessful()) {
                throw new IOException("HTTP " + response.code());
            }

            List<TsdOperationResponse> responses = response.body();
            if (responses == null) {
                responses = new ArrayList<>();
            }

            Map<String, TsdOperationResponse> byKey = new HashMap<>();
            for (TsdOperationResponse item : responses) {
                byKey.put(item.operationKey, item);
            }

            int applied = 0;
            int rejected = 0;
            int retried = 0;
            List<String> decisionMessages = new ArrayList<>();

            for (PendingOperation operation : operations) {
                TsdOperationResponse item = byKey.get(operation.operationKey);
                String operatorMessage = operatorMessage(item);
                String status = item == null ? null : item.status;
                if ("APPLIED".equals(status) || "ACCEPTED".equals(status) || "ALREADY_APPLIED".equals(status)) {
                    outbox.markSynced(operation.operationKey, operatorMessage);
                    applied++;
                    addDecisionMessage(item, operation, operatorMessage, decisionMessages);
                } else if ("REJECTED".equals(status)) {
                    outbox.markRejected(operation.operationKey, operatorMessage == null ? "Операция отклонена сервером" : operatorMessage);
                    rejected++;
                    addDecisionMessage(item, operation, operatorMessage, decisionMessages);
                } else if ("NEEDS_REVIEW".equals(status)) {
                    outbox.markRejected(operation.operationKey, operatorMessage == null ? "Операция требует разбора" : operatorMessage);
                    rejected++;
                    addDecisionMessage(item, operation, operatorMessage, decisionMessages);
                } else {
                    outbox.markRetry(operation.operationKey, operatorMessage == null ? "Нет ответа по операции" : operatorMessage);
                    retried++;
                    if ("receipt_close".equals(operation.operationType)) {
                        addDecisionMessage(item, operation, operatorMessage, decisionMessages);
                    }
                }
            }

            String summaryMessage;
            if (decisionMessages.isEmpty()) {
                summaryMessage = "Синхронизация завершена";
            } else {
                summaryMessage = "Синхронизация завершена. Решения: " + firstMessages(decisionMessages);
            }

            return new TsdSyncSummary(operations.size(), applied, rejected, retried, summaryMessage);
        } catch (Exception error) {
            for (PendingOperation operation : operations) {
                outbox.markRetry(operation.operationKey, error.getMessage() == null ? "Ошибка сети" : error.getMessage());
            }
            return new TsdSyncSummary(
                operations.size(),
                0,
                0,
                operations.size(),
                error.getMessage() == null ? "Ошибка синхронизации" : error.getMessage()
            );
        }
    }

    private TsdOperationRequest toRequest(PendingOperation operation) {
        return new TsdOperationRequest(deviceId, operation.operationKey, operation.operationType, operation.payload);
    }

    private String operatorMessage(TsdOperationResponse response) {
        if (response == null) {
            return null;
        }

        String text = response.resolutionMessage != null ? response.resolutionMessage : response.message;
        String reason = response.reviewReason == null ? null : reviewReasonLabel(response.reviewReason);
        if (reason != null && text != null) {
            return reason + ": " + text;
        }
        if (text != null) {
            return text;
        }
        return reason;
    }

    private void addDecisionMessage(
        TsdOperationResponse response,
        PendingOperation operation,
        String operatorMessage,
        List<String> messages
    ) {
        if (response == null || operatorMessage == null) {
            return;
        }
        // FIX: close results carry their actionable explanation in message.
        if (response.reviewReason == null && response.resolutionMessage == null
            && !"receipt_close".equals(operation.operationType)) {
            return;
        }

        messages.add(operation.operationType + ": " + operatorMessage);
    }

    private String reviewReasonLabel(String reason) {
        switch (reason) {
            case "INVENTORY_MISMATCH":
                return "Расхождение инвентаризации";
            case "SKU_NOT_FOUND":
                return "SKU не найден";
            case "BOX_NOT_FOUND":
                return "Короб не найден";
            case "RECEIPT_FAILED":
                return "Ошибка приемки";
            case "DEVICE_MISMATCH":
                return "Не тот ТСД";
            case "VALIDATION_ERROR":
                return "Ошибка данных";
            case "MANUAL_REJECT":
                return "Ручное отклонение";
            case "OTHER":
                return "Другая причина";
            default:
                return reason;
        }
    }

    private String firstMessages(List<String> messages) {
        StringBuilder builder = new StringBuilder();
        int limit = Math.min(2, messages.size());
        for (int index = 0; index < limit; index++) {
            if (index > 0) {
                builder.append("; ");
            }
            builder.append(messages.get(index));
        }
        return builder.toString();
    }
}

File: wms/apps/android-tsd/app/src/main/java/pro/logoff/wms/tsd/sync/TsdSyncRunner.kt
package pro.logoff.wms.tsd.sync

import pro.logoff.wms.tsd.data.OperationOutbox
import pro.logoff.wms.tsd.data.PendingOperation
import pro.logoff.wms.tsd.network.TsdOperationRequest
import pro.logoff.wms.tsd.network.TsdOperationResponse
import pro.logoff.wms.tsd.network.TsdSyncRequest
import pro.logoff.wms.tsd.network.WmsApi
import java.time.Instant
import kotlin.coroutines.cancellation.CancellationException

data class TsdSyncSummary(
    val sent: Int,
    val applied: Int,
    val rejected: Int,
    val retried: Int,
    val message: String,
)

class TsdSyncRunner(
    private val outbox: OperationOutbox,
    private val api: WmsApi,
    private val deviceId: String,
) {
    suspend fun syncPending(authorization: String): TsdSyncSummary {
        val operations = outbox.pending()
        if (operations.isEmpty()) {
            return TsdSyncSummary(
                sent = 0,
                applied = 0,
                rejected = 0,
                retried = 0,
                message = "Нет операций для синхронизации",
            )
        }

        return runCatching {
            val responses = api.syncOperations(
                authorization = authorization,
                request = TsdSyncRequest(
                    operations = operations.map { it.toRequest(deviceId) },
                    deviceClock = Instant.now().toString(),
                ),
            )
            val byKey = responses.associateBy { it.operationKey }
            var applied = 0
            var rejected = 0
            var retried = 0
            val decisionMessages = mutableListOf<String>()

            operations.forEach { operation ->
                val response = byKey[operation.operationKey]
                val operatorMessage = response?.operatorMessage()
                when (response?.status) {
                    "APPLIED", "ACCEPTED", "ALREADY_APPLIED" -> {
                        outbox.markSynced(operation.operationKey, operatorMessage)
                        applied += 1
                        response.addDecisionMessage(operation, operatorMessage, decisionMessages)
                    }

                    "REJECTED" -> {
                        outbox.markRejected(operation.operationKey, operatorMessage ?: "Операция отклонена сервером")
                        rejected += 1
                        response.addDecisionMessage(operation, operatorMessage, decisionMessages)
                    }

                    "NEEDS_REVIEW" -> {
                        outbox.markRejected(operation.operationKey, operatorMessage ?: "Операция требует разбора")
                        rejected += 1
                        response.addDecisionMessage(operation, operatorMessage, decisionMessages)
                    }

                    else -> {
                        // Русский комментарий: неизвестный или отсутствующий ответ оставляем в pending, чтобы повторить sync.
                        outbox.markRetry(operation.operationKey, operatorMessage ?: "Нет ответа по операции")
                        retried += 1
                    }
                }
            }

            val summaryMessage = if (decisionMessages.isEmpty()) {
                "Синхронизация завершена"
            } else {
                "Синхронизация завершена. Решения: ${decisionMessages.take(2).joinToString("; ")}"
            }

            TsdSyncSummary(
                sent = operations.size,
                applied = applied,
                rejected = rejected,
                retried = retried,
                message = summaryMessage,
            )
        }.getOrElse { error ->
            if (error is CancellationException) {
                throw error
            }
            operations.forEach { operation ->
                outbox.markRetry(operation.operationKey, error.message ?: "Ошибка сети")
            }
            TsdSyncSummary(
                sent = operations.size,
                applied = 0,
                rejected = 0,
                retried = operations.size,
                message = error.message ?: "Ошибка синхронизации",
            )
        }
    }

    private fun PendingOperation.toRequest(deviceId: String): TsdOperationRequest =
        TsdOperationRequest(
            deviceId = deviceId,
            operationKey = operationKey,
            operationType = operationType,
            payload = payload,
        )

    private fun TsdOperationResponse.operatorMessage(): String? {
        val text = resolutionMessage ?: message
        val reason = reviewReason?.let(::reviewReasonLabel)
        return when {
            reason != null && text != null -> "$reason: $text"
            text != null -> text
            reason != null -> reason
            else -> null
        }
    }

    private fun TsdOperationResponse.addDecisionMessage(
        operation: PendingOperation,
        operatorMessage: String?,
        messages: MutableList<String>,
    ) {
        if (operatorMessage == null || reviewReason == null && resolutionMessage == null) {
            return
        }

        messages += "${operation.operationType}: $operatorMessage"
    }

    private fun reviewReasonLabel(reason: String): String =
        when (reason) {
            "INVENTORY_MISMATCH" -> "Расхождение инвентаризации"
            "SKU_NOT_FOUND" -> "SKU не найден"
            "BOX_NOT_FOUND" -> "Короб не найден"
            "RECEIPT_FAILED" -> "Ошибка приемки"
            "DEVICE_MISMATCH" -> "Не тот ТСД"
            "VALIDATION_ERROR" -> "Ошибка данных"
            "MANUAL_REJECT" -> "Ручное отклонение"
            "OTHER" -> "Другая причина"
            else -> reason
        }
}
